serverless/cmd/client/client.go

```diff
@@ -0,0 +1,124 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package main is a read-only client command for interacting with serverless logs.
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian-examples/serverless/api"
+	"github.com/google/trillian-examples/serverless/internal/client"
+	"github.com/google/trillian/merkle/logverifier"
+	"github.com/google/trillian/merkle/rfc6962/hasher"
+)
+
+var (
+	storageURL   = flag.String("storage_url", "", "Log storage root URL, e.g. file:///path/to/log or https://log.server/and/path")
+	fromIndex    = flag.Uint64("from_index", 0, "Index for inclusion proof")
+	forEntryPath = flag.String("for_entry", "", "Path to entry for inclusion proof")
+)
+
+func main() {
+	flag.Parse()
+
+	rootURL, err := url.Parse(*storageURL)
+	if err != nil {
+		glog.Exitf("Invalid storage URL: %q", err)
+	}
+	f := newFetcher(rootURL)
+	logState, err := client.GetLogState(f)
+	if err != nil {
+		glog.Exitf("Failed to fetch log state: %q", err)
+	}
+
+	glog.Infof("Using tree size %d, with root 0x%0x", logState.Size, logState.RootHash)
+
+	args := flag.Args()
+	if len(args) == 0 {
+		glog.Exit("Please specify a command from [inclusion]")
+	}
+	switch args[0] {
+	case "inclusion":
+		err = inclusionProof(*logState, f, args[1:])
+	}
+	if err != nil {
+		glog.Exitf("Command %q failed: %q", args[0], err)
+	}
+}
+
+func inclusionProof(state api.LogState, f client.FetcherFunc, args []string) error {
+	entry, err := ioutil.ReadFile(*forEntryPath)
+	if err != nil {
+		return fmt.Errorf("failed to read entry from %q: %q", *forEntryPath, err)
+	}
+	h := hasher.DefaultHasher
+	lh := h.HashLeaf(entry)
+
+	builder, err := client.NewProofBuilder(state, h.HashChildren, f)
+	if err != nil {
+		return fmt.Errorf("failed to create proof builder: %w", err)
+	}
+
+	proof, err := builder.InclusionProof(*fromIndex)
+	if err != nil {
+		return fmt.Errorf("failed to get inclusion proof: %w", err)
+	}
+
+	lv := logverifier.New(hasher.DefaultHasher)
+	if err := lv.VerifyInclusionProof(int64(*fromIndex), int64(state.Size), proof, state.RootHash, lh); err != nil {
+		return fmt.Errorf("failed to verify inclusion proof: %q", err)
+	}
+
+	glog.Info("Inclusion verified.")
+	return nil
+}
+
+// newFetcher creates a fetched for the log at the given root location.
+func newFetcher(root *url.URL) client.FetcherFunc {
+	get := getByScheme[root.Scheme]
+	if get == nil {
+		panic(fmt.Errorf("unsupported URL scheme %s", root.Scheme))
+	}
+
+	return func(p string) ([]byte, error) {
+		u, err := root.Parse(p)
+		if err != nil {
+			return nil, err
+		}
+		return get(u)
+	}
+}
+
+var getByScheme = map[string]func(*url.URL) ([]byte, error){
```

Comment:
```suggestion
var getByScheme = map[string]func(*url.URL) ([]byte, error) {
```

I would have just made this a switch statement in `newFetcher` though?

Reply:
The space is not there 'cos it's an inline definition (gofmt removes it).

_Personally_, I think the switch-by-map is a little easier to read, but don't feel strongly either way:

```golang
// newFetcher creates a FetcherFunc for the log at the given root location.
func newFetcher(root *url.URL) client.FetcherFunc {
	var get func(*url.URL) ([]byte, error)
	switch root.Scheme {
	case "http", "https":
		get = readHTTP
	case "file":
		get = func(u *url.URL) ([]byte, error) {
			return ioutil.ReadFile(u.Path)
		}
	default:
		panic(fmt.Errorf("unsupported URL scheme %s", root.Scheme))
	}

	return func(p string) ([]byte, error) {
		u, err := root.Parse(p)
		if err != nil {
			return nil, err
		}
		return get(u)
	}
}
```

vs.

```golang
// newFetcher creates a FetcherFunc for the log at the given root location.
func newFetcher(root *url.URL) client.FetcherFunc {
	get := getByScheme[root.Scheme]
	if get == nil {
		panic(fmt.Errorf("unsupported URL scheme %s", root.Scheme))
	}

	return func(p string) ([]byte, error) {
		u, err := root.Parse(p)
		if err != nil {
			return nil, err
		}
		return get(u)
	}
}

var getByScheme = map[string]func(*url.URL) ([]byte, error){
	"http":  readHTTP,
	"https": readHTTP,
	"file": func(u *url.URL) ([]byte, error) {
		return ioutil.ReadFile(u.Path)
	},
}
```